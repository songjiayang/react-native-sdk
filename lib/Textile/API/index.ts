import { NativeModules } from 'react-native'
import { Buffer } from 'buffer'

import {
  File,
  ExternalInvite,
  ContactInfo,
  ContactInfoQueryResult,
  Overview,
  FileData,
  ThreadInfo,
  WalletAccount,
  BlockInfo,
  NotificationInfo,
  ThreadFilesInfo,
  ThreadFeedItem,
  ThreadMessageInfo,
  LogLevel
} from '../Models'
import {
  IMobilePreparedFiles,
  ICafeSession,
  ICafeSessions,
  IDirectory,
  MobilePreparedFiles,
  CafeSession,
  CafeSessions,
  Directory
} from '@textile/react-native-protobufs'

const { TextileNode } = NativeModules

/**
 * Returns the hash of the initial join block. Not the threadId of the final thread created/joined
 */
export async function acceptExternalThreadInvite(id_: string, key: string): Promise<string> {
  const result = await TextileNode.acceptExternalThreadInvite(id_, key) // returns hash
  return result as string
}

export async function acceptThreadInviteViaNotification(id_: string): Promise<string> {
  const result = await TextileNode.acceptThreadInviteViaNotification(id_) // returns addr
  return result as string
}

export async function addContact(contact: ContactInfo): Promise<void> {
  const contactJsonString = JSON.stringify(contact)
  await TextileNode.addContact(contactJsonString)
}

export async function addExternalThreadInvite(threadId: string): Promise<ExternalInvite> {
  const result = await TextileNode.addExternalThreadInvite(threadId)
  return JSON.parse(result) as ExternalInvite
}

export async function addSchema(jsonstr: string): Promise<File> {
  const result = await TextileNode.addSchema(jsonstr)
  return JSON.parse(result) as File
}

export async function addThread(key: string, name: string, shared: boolean): Promise<ThreadInfo> {
  const result = await TextileNode.addThread(key, name, shared)
  return JSON.parse(result) as ThreadInfo
}

export async function addThreadComment(blockId: string, body: string): Promise<string> {
  const result = await TextileNode.addThreadComment(blockId, body) // returns hash
  return result as string
}

export async function addThreadFiles(dir: IDirectory, threadId: string, caption?: string): Promise<BlockInfo> {
  const byteArray = Directory.encode(dir).finish()
  const buffer = Buffer.from(byteArray)
  const base64 = buffer.toString('base64')
  const result = await TextileNode.addThreadFiles(base64, threadId, caption)
  return JSON.parse(result) as BlockInfo
}

export async function addThreadFilesByTarget(target: string, threadId: string, caption?: string): Promise<BlockInfo> {
  const result = await TextileNode.addThreadFilesByTarget(target, threadId, caption)
  return JSON.parse(result) as BlockInfo
}

export async function addThreadIgnore(blockId: string): Promise<string> {
  const result = await TextileNode.addThreadIgnore(blockId) // returns hash
  return result as string
}

export async function addThreadInvite(threadId: string, inviteeId: string): Promise<string> {
  const result = await TextileNode.addThreadInvite(threadId, inviteeId) // returns hash
  return result as string
}

export async function addThreadLike(blockId: string): Promise<string> {
  const result = await TextileNode.addThreadLike(blockId) // returns hash
  return result as string
}

export async function addThreadMessage(threadId: string, body: string): Promise<string> {
  const result = await TextileNode.addThreadMessage(threadId, body) // returns hash
  return result as string
}

export async function address(): Promise<string> {
  const result = await TextileNode.address()
  return result as string
}

export async function avatar(): Promise<string | undefined> {
  const result: string = await TextileNode.avatar()
  return result.length > 0 ? result : undefined
}

export async function cafeSession(peerId: string): Promise<ICafeSession | undefined> {
  const result = await TextileNode.cafeSession(peerId)
  if (!result) {
    return undefined
  }
  const buffer = Buffer.from(result, 'base64')
  return CafeSession.decode(buffer)
}

export async function cafeSessions(): Promise<ICafeSessions | undefined> {
  const result = await TextileNode.cafeSessions()
  if (!result) {
    return undefined
  }
  const buffer = Buffer.from(result, 'base64')
  return CafeSessions.decode(buffer)
}

export async function checkCafeMessages(): Promise<void> {
  await TextileNode.checkCafeMessages()
}

export async function contact(id_: string): Promise<ContactInfo> {
  const result = await TextileNode.contact(id_)
  return JSON.parse(result) as ContactInfo
}

export async function contactThreads(id_: string): Promise<ReadonlyArray<ThreadInfo>> {
  const result = await TextileNode.contactThreads(id_)
  return JSON.parse(result) as ReadonlyArray<ThreadInfo>
}

export async function contacts(): Promise<ReadonlyArray<ContactInfo>> {
  const result = await TextileNode.contacts()
  return JSON.parse(result) as ReadonlyArray<ContactInfo>
}

export async function countUnreadNotifications(): Promise<number> {
  const result = await TextileNode.countUnreadNotifications()
  return result as number
}

export async function deregisterCafe(peerId: string): Promise<void> {
  await TextileNode.deregisterCafe(peerId)
}

export async function fileData(hash: string): Promise<FileData> {
  const result = await TextileNode.fileData(hash)
  return JSON.parse(result) as FileData
}

export async function findContact(username: string, limit: number, wait: number): Promise<ContactInfoQueryResult> {
  const result = await TextileNode.findContact(username, limit, wait)
  return JSON.parse(result) as ContactInfoQueryResult
}

export async function ignoreThreadInviteViaNotification(id_: string): Promise<string> {
  const result = await TextileNode.ignoreThreadInviteViaNotification(id_)
  return result as string
}

// Note: pth is <target>/<index>, e.g., "Qm.../0"
export async function imageFileDataForMinWidth(pth: string, minWidth: number): Promise<FileData> {
  const result = await TextileNode.imageFileDataForMinWidth(pth, minWidth)
  return JSON.parse(result) as FileData
}

export async function notifications(offset: string, limit: number): Promise<ReadonlyArray<NotificationInfo>> {
  const result = await TextileNode.notifications(offset, limit)
  return JSON.parse(result) as ReadonlyArray<NotificationInfo>
}

export async function overview(): Promise<Overview> {
  const result = await TextileNode.overview()
  return JSON.parse(result) as Overview
}

export async function peerId(): Promise<string> {
  const result = await TextileNode.peerId()
  return result as string
}

export async function prepareFiles(path: string, threadId: string): Promise<IMobilePreparedFiles> {
  const result = await TextileNode.prepareFiles(path, threadId)
  const buffer = Buffer.from(result, 'base64')
  return MobilePreparedFiles.decode(buffer)
}

export async function prepareFilesAsync(path: string, threadId: string): Promise<IMobilePreparedFiles> {
  const result = await TextileNode.prepareFilesAsync(path, threadId)
  const buffer = Buffer.from(result, 'base64')
  return MobilePreparedFiles.decode(buffer)
}

export async function profile(): Promise<ContactInfo> {
  const result = await TextileNode.profile()
  return JSON.parse(result) as ContactInfo
}

export async function readAllNotifications(): Promise<void> {
  await TextileNode.readAllNotifications()
}

export async function readNotification(id_: string): Promise<void> {
  await TextileNode.readNotification(id_)
}

export async function refreshCafeSession(peerId: string): Promise<ICafeSession | undefined> {
  const result = await TextileNode.refreshCafeSession(peerId)
  if (!result) {
    return undefined
  }
  const buffer = Buffer.from(result, 'base64')
  return CafeSession.decode(buffer)
}

export async function registerCafe(peerId: string, token: string): Promise<void> {
  await TextileNode.registerCafe(peerId, token)
}

export async function removeThread(id_: string): Promise<string> {
  const result = await TextileNode.removeThread(id_) // returns hash b58 string
  return result as string
}

export async function seed(): Promise<string> {
  const result = await TextileNode.seed()
  return result as string
}

export async function setAvatar(id_: string): Promise<void> {
  await TextileNode.setAvatar(id_)
}

export async function setLogLevels(levels: Map<string, LogLevel>): Promise<void> {
  await TextileNode.setLogLevels(levels)
}

export async function setUsername(username: string): Promise<void> {
  await TextileNode.setUsername(username)
}

export async function start(): Promise<void> {
  await TextileNode.start()
}

export async function stop(): Promise<void> {
  await TextileNode.stop()
}

export async function threadFeed(offset: string, limit: number, threadId?: string): Promise<ReadonlyArray<ThreadFeedItem>> {
  const result = await TextileNode.threadFeed(offset, limit, threadId)
  return JSON.parse(result) as ReadonlyArray<ThreadFeedItem>
}

export async function threadFiles(offset: string, limit: number, threadId?: string): Promise<ReadonlyArray<ThreadFilesInfo>> {
  const result = await TextileNode.threadFiles(offset, limit, threadId)
  return JSON.parse(result) as ReadonlyArray<ThreadFilesInfo>
}

export async function threadMessages(offset: string, limit: number, threadId?: string): Promise<ReadonlyArray<ThreadMessageInfo>> {
  const result = await TextileNode.threadMessages(offset, limit, threadId)
  return JSON.parse(result) as ReadonlyArray<ThreadMessageInfo>
}

export async function threadInfo(threadId: string): Promise<ThreadInfo> {
  const result = await TextileNode.threadInfo(threadId)
  return JSON.parse(result) as ThreadInfo
}

export async function threads(): Promise<ReadonlyArray<ThreadInfo>> {
  const result = await TextileNode.threads()
  return JSON.parse(result) as ReadonlyArray<ThreadInfo>
}

export async function username(): Promise<string | undefined> {
  const result: string = await TextileNode.username()
  return result.length > 0 ? result : undefined
}

export async function version(): Promise<string> {
  const result = await TextileNode.version()
  return result as string
}

// Order of things to init and create the repo:
// MobileNewTextile If error, inspect it and run next steps or migration
// MobileNewWallet returns recovery phrase
// MobileWalletAccountAt returns seed and address
// MobileInitRepo only run one time ever
// MobileNewTextile

export async function initRepo(seed: string, repoPath: string, logToDisk: boolean, debug: boolean): Promise<void> {
  return await TextileNode.initRepo(seed, repoPath, logToDisk, debug)
}

export async function migrateRepo(repoPath: string): Promise<void> {
  await TextileNode.migrateRepo(repoPath)
}

export async function newTextile(repoPath: string, debug: boolean): Promise<void> {
  await TextileNode.newTextile(repoPath, debug)
}

export async function newWallet(wordCount: number): Promise<string> {
  const result = await TextileNode.newWallet(wordCount)
  return result as string
}

export async function walletAccountAt(phrase: string, index: number, password?: string): Promise<WalletAccount> {
  const result = await TextileNode.walletAccountAt(phrase, index, password) // return seed and address
  return JSON.parse(result) as WalletAccount
}
