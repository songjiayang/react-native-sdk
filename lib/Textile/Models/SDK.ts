import { AppStateStatus } from 'react-native'
import { pb } from './pb'

export interface ContactSearchResult {
  type: 'result'
  contact: pb.IContact,
  local: boolean
}

export interface ContactSearchError {
  type: 'error'
  error: string
}

export interface ContactSearchComplete {
  type: 'complete'
}

export type ContactSearchEvent = ContactSearchResult | ContactSearchError | ContactSearchComplete

export interface BufferJSON {
  buffer: string
}

export interface CafeConfig {
  TEXTILE_CAFE_GATEWAY_URL: string
  TEXTILE_CAFE_TOKEN: string
  TEXTILE_CAFE_OVERRIDE?: string
}

export enum NodeState {
  'nonexistent' = 'nonexistent',
  'creating' = 'creating',
  'created' = 'created', // Node has been created, on it's way to starting
  'starting' = 'starting',
  'started' = 'started',
  'stopping' = 'stopping',
  'stopped' = 'stopped', // Node has been explicitly stopped, different than created
  'creatingWallet' = 'creatingWallet',
  'derivingAccount' = 'derivingAccount',
  'initializingRepo' = 'initializingRepo',
  'walletInitSuccess' = 'walletInitSuccess',
  'postMigration' = 'postMigration'
}

export enum SchemaType {
  MEDIA = 'MEDIA',
  CAMERA_ROLL = 'CAMERA_ROLL',
  JSON = 'JSON'
}
export interface StoredNodeState {
  state: NodeState
  error?: string
}

export type TextileAppStateStatus = AppStateStatus | 'unknown' | 'backgroundFromForeground'

// TODO: remove capitalization
export interface TextileConfig {
  RELEASE_TYPE?: string
  SELF_MANAGE_APP_STATE?: boolean
  MINIMUM_SLEEP_MINUTES?: number
  RUN_BACKGROUND_TASK?: () => boolean
}

export interface TextileOptions {
  debug?: boolean
}
