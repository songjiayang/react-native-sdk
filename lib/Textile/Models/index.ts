export * from './API'
export * from './SDK'
