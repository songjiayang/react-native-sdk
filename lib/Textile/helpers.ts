
export function getHMS() {
  const now = new Date()
  return [
    now.getHours().toString(),
    now.getMinutes().toString(),
    now.getSeconds().toString()
  ].join(':')
}

export function createTimeout(ms: number, promise: Promise<any>): Promise<any> {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      reject(new Error('timeout'))
    }, ms)
    promise.then(resolve, reject)
  })
}

export function delay(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve()
    }, ms)
  })
}
