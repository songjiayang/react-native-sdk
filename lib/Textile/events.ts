import {
  DeviceEventEmitter
} from 'react-native'
import { NodeState, TextileAppStateStatus } from './Models'

export const keys = {
  newNodeState: '@textile/newNodeState',
  createAndStartNode: '@textile/createAndStartNode',
  startNodeFinished: '@textile/startNodeFinished',
  stopNodeAfterDelayStarting: '@textile/stopNodeAfterDelayStarting',
  stopNodeAfterDelayCancelled: '@textile/stopNodeAfterDelayCancelled',
  stopNodeAfterDelayFinishing: '@textile/stopNodeAfterDelayFinishing',
  stopNodeAfterDelayComplete: '@textile/stopNodeAfterDelayComplete',
  appStateChange: '@textile/appStateChange',
  updateProfile: '@textile/updateProfile',
  newErrorMessage: '@textile/newErrorMessage',
  appNextState: '@textile/appNextState',
  migrationNeeded: '@textile/migrationNeeded',
  setRecoveryPhrase: '@textile/setRecoveryPhrase',
  walletInitSuccess: '@textile/walletInitSuccess',
  backgroundTask: '@textile/backgroundTask',
  error: '@textile/error'
}

export function newError(message: string, type: string) {
  DeviceEventEmitter.emit(keys.error, {type, message})
}

export function nonInitializedError() {
  newError('nonInitializedError', 'Error: Attempt to use a Textile method reserved for an initialized instance.')
}
export function backgroundTask () {
  DeviceEventEmitter.emit(keys.backgroundTask)
}

export function newNodeState (state: NodeState) {
  DeviceEventEmitter.emit(keys.newNodeState, {state})
}
export function createAndStartNode () {
  DeviceEventEmitter.emit(keys.createAndStartNode)
}

export function startNodeFinished () {
  DeviceEventEmitter.emit(keys.startNodeFinished)
}

export function stopNodeAfterDelayStarting () {
  DeviceEventEmitter.emit(keys.stopNodeAfterDelayStarting)
}

export function stopNodeAfterDelayCancelled () {
  DeviceEventEmitter.emit(keys.stopNodeAfterDelayCancelled)
}

export function stopNodeAfterDelayFinishing () {
  DeviceEventEmitter.emit(keys.stopNodeAfterDelayFinishing)
}

export function stopNodeAfterDelayComplete () {
  DeviceEventEmitter.emit(keys.stopNodeAfterDelayComplete)
}
export function appStateChange (previousState: TextileAppStateStatus, newState: TextileAppStateStatus) {
  DeviceEventEmitter.emit(keys.appStateChange, {previousState: previousState as string, newState: newState as string})
}

export function updateProfile () {
  DeviceEventEmitter.emit(keys.updateProfile)
}
export function walletInitSuccess () {
  DeviceEventEmitter.emit(keys.walletInitSuccess)
}

export function setRecoveryPhrase (recoveryPhrase: string) {
  DeviceEventEmitter.emit(keys.setRecoveryPhrase, {recoveryPhrase})
}
export function migrationNeeded () {
  DeviceEventEmitter.emit(keys.migrationNeeded)
}

export function appNextState (nextState: string) {
  DeviceEventEmitter.emit(keys.appNextState, {nextState})
}
