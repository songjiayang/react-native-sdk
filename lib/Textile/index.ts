import { AppState, AppStateStatus, DeviceEventEmitter } from 'react-native'
import {
  DiscoveredCafes,
  TextileAppStateStatus,
  TextileOptions,
  WalletAccount,
  NodeState,
  TextileConfig
} from './Models'
import * as API from './API'
import TextileStore from './store'
import Events from '../Events'
import TextileMigration from './migration'
import * as TextileEvents from './events'
import { createTimeout, delay } from './helpers'
import BackgroundTimer from 'react-native-background-timer'
import BackgroundFetch from 'react-native-background-fetch'
import RNFS from 'react-native-fs'

import { ICafeSessions, ICafeSession } from '@textile/react-native-protobufs'

const packageFile = require('./../../package.json')
export const VERSION = packageFile.version

const MIGRATION_NEEDED_ERROR = 'repo needs migration'
const INIT_NEEDED_ERROR = 'repo does not exist, initialization is required'

class Textile {
  // Temp instance of the app's redux store while I remove deps to it
  api: any
  migration = new TextileMigration()
  _debug = false
  _store = new TextileStore()
  _nativeEvents = Events
  _config: TextileConfig = {
    RELEASE_TYPE: 'development'
  }
  _initialized = false

  repoPath = `${RNFS.DocumentDirectoryPath}/textile-go`

  constructor(options: TextileOptions) {
    if (options.debug) {
      this._debug = true
    }
    this.api = API
    console.info('Initializing @textile/react-native-sdk v. ' + VERSION)
  }

  /* ---- Functions to wire into app ------ */
  backgroundFetch () {
    this.startBackgroundTask()
  }
  locationUpdate () {
    this.startBackgroundTask()
  }

  // De-register the listeners
  tearDown() {
    // Clear on out too if detected to help speed up any startup time
    // Clear all our listeners
    this._nativeEvents.removeAllListeners()
    // TODO: be sure to limit to only internal listeners (same above)
    DeviceEventEmitter.removeAllListeners()
    if (!this._config.SELF_MANAGE_APP_STATE) {
      AppState.removeEventListener('change', (nextState: AppStateStatus) => {
        TextileEvents.appNextState(nextState)
        this.nextAppState(nextState)
      })
    }
  }

  // setup should only be run where the class will remain persistent so that
  // listeners will be wired in to one instance only,
  setup(config?: TextileConfig) {
    // if config provided, set it
    if (config) {
      this._config = config
    }
    // Clear storage to fresh state
    this._store.clear()
    // Clear state on setup
    // Setup our within sdk listeners
    this._nativeEvents.addListener('onOnline', () => {
      this._store.setNodeOnline(true)
    })

    DeviceEventEmitter.addListener('@textile/createAndStartNode', (payload) => {
      this.createAndStartNode()
    })

    if (!this._config.SELF_MANAGE_APP_STATE) {
      // SDK automatically detects app state changes manages the node
      AppState.addEventListener('change', (nextState: AppStateStatus) => {
        TextileEvents.appNextState(nextState)
        this.nextAppState(nextState)
      })
    } else {
      // Alternatively, the developer can trigger changes manually via an notifyAppStateChange event
      DeviceEventEmitter.addListener('@textile/notifyAppStateChange', (payload) => {
        if (!payload || !payload.nextState) {
          return
        }
        TextileEvents.appNextState(payload.nextState)
        this.nextAppState(payload.nextState)
      })
    }

    this.initializeAppState()

    this._initialized = true
  }

  isInitializedCheck = () => {
    if (!this._initialized) {
      TextileEvents.nonInitializedError()
      throw new Error('Attempt to call a Textile instance method on an uninitialized instance')
    }
  }

  /* ---- STATE BASED METHODS ----- */
  //  All methods here should only be called as the result of a sequenced kicked off
  //  By an event and detected by the persistent instance that executed setup()

  initializeAppState = async () => {
    const defaultAppState = 'default' as TextileAppStateStatus
    // if for some reason initialized will ever be called from a non-blank state (?), we need the below
    // const storedState = await this._store.getAppState()
    // let defaultAppState = 'default' as TextileAppStateStatus
    // if (storedState) {
    //   defaultAppState = JSON.parse(storedState) as TextileAppStateStatus
    // }
    // wait just a moment in case we beat native state
    await delay(10)
    const currentAppState = AppState.currentState
    const queriedAppState = currentAppState || 'unknown'
    await this.appStateChange(defaultAppState, queriedAppState)
  }

  startBackgroundTask = async () => {
    const shouldRun = await this.shouldRunBackgroundTask()
    if (!shouldRun) {
      return
    }
    await this._store.setLastBackgroundEvent()
    const currentState = await this.appState()
    // const currentState = yield select(TextileNodeSelectors.appState)
    // ensure we don't cause things in foreground
    if (currentState === 'background' || currentState === 'backgroundFromForeground') {
      await this.appStateChange(currentState, 'background')
    }
  }

  // Simply create the node, useful only if you want to create in advance of starting
  createNode = async () => {
    this.isInitializedCheck()
    const debug = this._config.RELEASE_TYPE !== 'production'
    await this.updateNodeState(NodeState.creating)
    const needsMigration = await this.migration.requiresFileMigration(this.repoPath)
    if (needsMigration) {
      await this.migration.runFileMigration(this.repoPath)
    }
    await this.api.newTextile(this.repoPath, debug)

    await this.updateNodeState(NodeState.created)
  }

  // Start the node, create it if it doesn't exist. Safe to call on every start.
  createAndStartNode = async () => {
    // TODO
    /* In redux/saga world, we did a // yield call(() => task.done) to ensure this wasn't called
    while already running. Do we need the same check to ensure it doesn't happen here?
    */
    this.isInitializedCheck()
    try {

      await this.createNode()

      await this.updateNodeState(NodeState.starting)

      await this.api.start()

      const sessions: ICafeSessions = await this.api.cafeSessions()
      if (!sessions || !sessions.values || sessions.values.length < 1) {
        const cafeOverride = this._config.TEXTILE_CAFE_OVERRIDE
        if (cafeOverride) {
          await this.api.registerCafe(cafeOverride as string)
        } else if (this._config.TEXTILE_CAFE_GATEWAY_URL) {
          await this.discoverAndRegisterCafes()
        }
      }
      await this.updateNodeState(NodeState.started)
      TextileEvents.startNodeFinished()
    } catch (error) {
      try {
        if (error.message === MIGRATION_NEEDED_ERROR) {
          // instruct the node to export data to files
          await this.api.migrateRepo(this.repoPath)
          // store the fact there is a pending migration in the preferences redux persisted state
          TextileEvents.migrationNeeded()
          // call the create/start sequence again
          TextileEvents.createAndStartNode()
        } else if (error.message === INIT_NEEDED_ERROR) {
          await this.updateNodeState(NodeState.creatingWallet)
          const recoveryPhrase: string = await this.api.newWallet(12)
          TextileEvents.setRecoveryPhrase(recoveryPhrase)
          await this.updateNodeState(NodeState.derivingAccount)
          const walletAccount: WalletAccount = await this.api.walletAccountAt(recoveryPhrase, 0)
          await this.updateNodeState(NodeState.initializingRepo)
          await this.api.initRepo(walletAccount.seed, this.repoPath, true, debug)
          TextileEvents.createAndStartNode()
          TextileEvents.walletInitSuccess()
        } else {
          await this.updateNodeStateError(error)
        }
      } catch (error) {
        await this.updateNodeStateError(error)
      }
    }
  }

  // Useful if an app wishes to shut down the node
  shutDown = async () => {
    await this.stopNode()
  }

  // Primarily an internal function
  manageNode = async (previousState: TextileAppStateStatus, newState: TextileAppStateStatus) => {
    this.isInitializedCheck()
    if (newState === 'active' || newState === 'background' || newState === 'backgroundFromForeground') {
      await TextileEvents.appStateChange(previousState, newState)
    }
    if (newState === 'active' || newState === 'background') {
      this.createAndStartNode()
    }
    if (newState === 'background' || newState === 'backgroundFromForeground') {
      await this.backgroundTaskRace()
    }
  }

  discoverAndRegisterCafes = async () => {
    this.initializeAppState()
    try {
      const cafes = await createTimeout(10000, this.discoverCafes())
      const discoveredCafes = cafes as DiscoveredCafes
      await this.api.registerCafe(discoveredCafes.primary.url)
      await this.api.registerCafe(discoveredCafes.secondary.url)
    } catch (error) {
      // When this happens, you should retry the discover and register...
      TextileEvents.newError('cafeDiscoveryError', 'cafe discovery timed out, internet connection needed')
    }
  }

  /* ----- STATE FREE PUBLIC SELECTORS ----- */
  isInitialized = () => {
    return this._initialized
  }

  appState = async (): Promise<TextileAppStateStatus> => {
    const storedState = await this._store.getAppState()
    const currentState = storedState || 'unknown' as TextileAppStateStatus
    return currentState
  }

  nodeOnline = async (): Promise<boolean> => {
    const online = await this._store.getNodeOnline()
    return !!online // store can return void, in which case default return false
  }

  nodeState = async (): Promise<NodeState> => {
    const storedState = await this._store.getNodeState()
    if (!storedState) {
      return NodeState.nonexistent
    }
    return storedState.state
  }

  // Client should use this once account is onboarded to register with Cafe
  getCafeSessions = async (): Promise<ReadonlyArray<ICafeSession>> => {
    const sessions: Readonly<ICafeSessions> = await this.api.cafeSessions()
    if (!sessions) {
      return []
    }
    const values: ReadonlyArray<ICafeSession> | undefined | null = sessions.values
    if (!values) {
      return []
    } else {
      return values
    }
  }

  // Client should use this if cafe sessions are detected as expired
  getRefreshedCafeSessions = async (): Promise<ReadonlyArray<ICafeSession>> => {
    const sessions: Readonly<ICafeSessions> = await this.api.cafeSessions()
    if (!sessions) {
      return []
    }
    const values: ReadonlyArray<ICafeSession> | undefined | null = sessions.values
    if (!values) {
      return []
    } else {
      const refreshedValues: ReadonlyArray<ICafeSession> = await Promise.all(
        values.map(async (session) => await this.api.refreshCafeSession(session.id!))
      )
      return refreshedValues
    }
  }

  /* ------ INTERNAL METHODS ----- */
  private shouldRunBackgroundTask = async (): Promise<boolean> => {
    const MINIMUM_MINUTES_BETWEEN_TASKS = 10
    const now = Number((new Date()).getTime())
    const last = await this._store.getLastBackgroundEvent()
    // previous time set and set too recently
    if (last && (now - last) < 1000 * 60 * MINIMUM_MINUTES_BETWEEN_TASKS) {
      return false
    }
    return true
  }

  private discoverCafes = async () => {
    if (!this._initialized) {
      TextileEvents.nonInitializedError()
      return
    }
    const response = await fetch(`${this._config.TEXTILE_CAFE_GATEWAY_URL}/cafes`, { method: 'GET' })
    if (response.status < 200 || response.status > 299) {
      throw new Error(`Status code error: ${response.statusText}`)
    }
    const discoveredCafes = await response.json() as DiscoveredCafes
    return discoveredCafes
  }

  private updateNodeStateError = async (error: Error) => {
    const storedState = await this._store.getNodeState()
    const state = storedState ? storedState.state : NodeState.nonexistent
    await this._store.setNodeState({state, error: error.message})
  }
  private nextAppState = async (nextState: TextileAppStateStatus) => {
    this.isInitializedCheck()
    const previousState = await this.appState()
        // const currentState = this.store.getState().textileNode.appState
    const newState: TextileAppStateStatus = nextState === 'background' && (previousState === 'active' || previousState === 'inactive') ? 'backgroundFromForeground' : nextState
    if (newState !== previousState || newState === 'background') {
      await this.appStateChange(previousState, newState)
    }
  }
  private appStateChange = async (previousState: TextileAppStateStatus, nextState: TextileAppStateStatus) => {
    this.isInitializedCheck()
    await this._store.setAppState(nextState)
    await this.manageNode(previousState, nextState)
  }
  private updateNodeState = async (state: NodeState) => {
    this.isInitializedCheck()
    await this._store.setNodeState({state})
    TextileEvents.newNodeState(state)
  }

  /* ----- PRIVATE - EVENT EMITTERS ----- */
  private stopNode = async () => {
    await this.updateNodeState(NodeState.stopping)
    await this.api.stop()
    this._store.setNodeOnline(false)
    await this.updateNodeState(NodeState.stopped)
  }

  private backgroundTaskRace = async () => {
    // This race cancels whichever effect looses the race, so a foreground event will cancel stopping the node
    //
    // Using the race effect, if we get a foreground event while we're waiting
    // to stop the node, cancel the stop and let it keep running
    await BackgroundTimer.start()
    const ms = 20000
    let cancelled = false

    const foregroundEvent = DeviceEventEmitter.addListener('@textile/appNextState', (payload) => {
      if (payload.nextState === 'active' && !cancelled) {
        TextileEvents.stopNodeAfterDelayCancelled()
        cancelled = true
      }
    })

    cancelSequence:
    while (!cancelled) {
        TextileEvents.stopNodeAfterDelayStarting()
        await this.api.checkCafeMessages() // do a quick check for new messages
        await delay(ms / 2)
        if (cancelled) { // cancelled by event, so abort sequence
          foregroundEvent.remove() // remove our event listener
          break cancelSequence
        }
        await this.api.checkCafeMessages()
        await delay(ms / 2)
        if (cancelled) { // cancelled by event, so abort sequence
          foregroundEvent.remove() // remove our event listener
          break cancelSequence
        }
        // enter stopping sequence
        foregroundEvent.remove() // remove our event listener
        TextileEvents.stopNodeAfterDelayFinishing()
        await this.stopNode() // stop the node
        cancelled = true // be sure to exit the loop
    }
    await BackgroundTimer.stop()
    // TODO: this might be better in a client provided callback
    await BackgroundFetch.finish(BackgroundFetch.FETCH_RESULT_NEW_DATA)
  }
}

export default new Textile({})
