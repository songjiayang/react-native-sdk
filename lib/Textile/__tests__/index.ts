import Textile from '../'
import { NodeState } from '../Models'
import { delay } from '../helpers'

describe('rn textile', () => {
  describe('stateless functions', () => {
    it('selectors', async () => {
      expect(Textile.appState()).resolves.toEqual('unknown')
      expect(Textile.nodeOnline()).resolves.toEqual(false)
      expect(Textile.nodeState()).resolves.toEqual(NodeState.nonexistent)
    })
  })
  describe('state functions', () => {
    it('startup sequence', async () => {
      expect(Textile.isInitialized()).toEqual(false)
      expect(Textile.setup()).toMatchSnapshot()
      expect(Textile.isInitialized()).toEqual(true)
      await delay(50)
      expect(Textile.appState()).resolves.toEqual('active')
    })
  })
})
