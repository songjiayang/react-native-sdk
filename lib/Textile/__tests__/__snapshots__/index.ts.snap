// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rn textile state functions startup sequence 1`] = `undefined`;
